### PR TITLE
Refactor and fix SDMA driver

### PR DESCRIPTION
**Pull request by dbaluta**
Our SDMA drivers has problems when running stress tests. What happens is that channels are not scheduled when running subsequent aplay/arecord commands.

This patch series aims to simplify to code and take a new approach on scheduling the channels. RM says that a condition for a channel to be runnable is      (HE[i] or HO[i]) and (EP[i] or EO[i])

So, foar we have overriden the manual starting of the channel by setting HO[i] = 0, but this creates problems. In order to have a better control on scheduling the channels we use both conditions to schedule the channel:
* make HO[i] =  0
* make EO[i] =  0

With this patchseires and also changing sdma_watermark to half is current value stress test work fine for 1000 iterations.

This means that a channel will run when both the channel was manually started and a DMA request event arrived for the corresponding channel.

### CODE REVIEW

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:95]: Does this actually have an effect? From what I recall from documentation, writing a 0 does NOT clear the bit but rather it leaves it unchanged, and writing a 1 does the actual action of setting the actual internal start bit. I think the old code is the correct one and this one could accidentally start other channels more times than required.

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:100]: Accidentally stop multiple channels? Same explanation as previous commit, you are stopping ALL active channels except those that have a second "start" command given to them.

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:440]: Didn't I initialize the channel indices at driver probe time? Still, good to have, thanks.

**Review by paulstelian97 — CHANGES_REQUESTED** [review]
Blocking until you've re-read the documentation of the SDMA_HSTART and SDMA_HSTOP_STAT bits and either explain why dma_reg_update_bits is correct for them or change back to dma_reg_write.

File: src/drivers/imx/sdma.c
Comment (dbaluta) [reply src/drivers/imx/sdma.c:100]: @paulstelian97 I think you are right. Writing a 1 will stop a started channel. Will update. Same for enable_channel.

**Comment by dbaluta** [comment]
@paulstelian97 comments updated plus introduced a patch that updated the SDMA watermark to a working value.

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:575]: Commit message has a typo on this one:

"In order, for the channel to be reschedule we need to start again the
HSTART_HE after we got an interrupt that a b_o_d was processed."

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [inline src/drivers/imx/sdma.c:776]: Where does each of multiplying by 4 and dividing by 8 come from? I feel like at least the multiplication by 4 can be replaced with a multiplication by the width?

Comment (dbaluta) [reply src/drivers/imx/sdma.c:776]: @paulstelian97 yes 4 comes from width and i will replace it. Then / 8 comes from multiple run tests, where this was the best option.

Ideally, as SDMA owner told me it should work using the same watermark as the DAI.  I will still investigate this until we completely understand what happens. 

So, lets just keep this PR open for few more days.

File: src/drivers/imx/sdma.c
Comment (dbaluta) [reply src/drivers/imx/sdma.c:575]: @paulstelian97 thanks. will fix.

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [reply src/drivers/imx/sdma.c:776]: The main thing is, the SDMA watermark must divide the size of the BD otherwise we get unexpected behavior that is harder to quantify. Does this ensure that?

Comment (dbaluta) [reply src/drivers/imx/sdma.c:776]: @paulstelian97 yes, that's ensure that. And I choose this values explicitly to validate the condition. The thing is that the documentation doesn't explicitly mention about watermark to divide the BD count :).

Comment (paulstelian97) [reply src/drivers/imx/sdma.c:776]: Well it technically doesn't _break_ if you don't ensure the condition, but the behavior we're getting is weird because at the end of the BD it could definitely transfer less than what the watermark says.

File: src/drivers/imx/sdma.c
Comment (paulstelian97) [reply src/drivers/imx/sdma.c:575]: Waiting for this fix, I guess I can approve after that.

Comment (dbaluta) [reply src/drivers/imx/sdma.c:575]: Oh, sorry forgot about this. Pushed now.

File: src/drivers/imx/sdma.c
Comment (dbaluta) [reply src/drivers/imx/sdma.c:776]: Indeed, these indeed are fine tuning patches. That we will surely do in the future. But lets now have an SDMA driver which is stable and gets work done.

Comment (paulstelian97) [reply src/drivers/imx/sdma.c:776]: `width`?

Comment (dbaluta) [reply src/drivers/imx/sdma.c]: oh, thanks for noticing that. I thought that Ive already fixed that. Should be ok now.

**Review by paulstelian97 — APPROVED** [review]

**Comment by dbaluta** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
